### PR TITLE
feat(bridge-history): change FinalizeBatch event logic

### PR DESCRIPTION
**Pull request by amoylan2**
### Purpose or design rationale of this PR

Upgrade the logic when dealing with FinalizeBatch event. 
This is intended to support darwin as well as backward compatible.

One thing need caution: "GetLastFinalizedBatchIndex" from db can not use any index and it's not suitable to add such index since there will be almost 100% batches are all finalized as time goes. Do we need add a key-value mode cache to save the "last finalized batch index"? it can be a small kv table in postgres or using redis..

### PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [ ] build: Changes that affect the build system or external dependencies (example scopes: yarn, eslint, typescript)
- [ ] ci: Changes to our CI configuration files and scripts (example scopes: vercel, github, cypress)
- [ ] docs: Documentation-only changes
- [x] feat: A new feature
- [ ] fix: A bug fix
- [ ] perf: A code change that improves performance
- [ ] refactor: A code change that doesn't fix a bug, or add a feature, or improves performance
- [ ] style: Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
- [ ] test: Adding missing tests or correcting existing tests

### Deployment tag versioning

Has `tag` in `common/version.go` been updated or have you added `bump-version` label to this PR?

- [ ] No, this PR doesn't involve a new deployment, git tag, docker image tag
- [ ] Yes

### Breaking change label

Does this PR have the `breaking-change` label?

- [x] No, this PR is not a breaking change
- [ ] Yes

### CODE REVIEW

File: bridge-history-api/internal/orm/batch_event.go [hunk 1]
```diff
@@ -89,8 +89,9 @@ func (c *BatchEvent) InsertOrUpdateBatchEvents(ctx context.Context, l1BatchEvent
 				return fmt.Errorf("failed to insert or ignore batch event, error: %w", err)
 			}
 		case btypes.BatchStatusTypeFinalized:
+			// After darwin, FinalizeBatch event signals a range of batches are finalized,
+			// thus losing the batch hash info. Meanwhile, only batch_index is enough to find the target batch.
```
Comment (colinlyguo) [inline bridge-history-api/internal/orm/batch_event.go]:
```suggestion
			// thus losing the batch hash info. Meanwhile, only batch_index is enough to find the target batch when finalizing batches.
```

**Comment by colinlyguo** [comment]
This feature can be directly merged into the develop branch since it's backward compatible, we can deploy it first.

For discussion about: _One thing need caution: "GetLastFinalizedBatchIndex" from db can not use any index and it's not suitable to add such index since there will be almost 100% batches are all finalized as time goes. Do we need add a key-value mode cache to save the "last finalized batch index"? it can be a small kv table in postgres or using redis.._

this can be mitigated by the "composite index" as follows:
```
CREATE INDEX IF NOT EXISTS idx_be_batch_index_batch_status ON batch_event_v2 (batch_index, batch_status);
```

but I think using `batch_index` is sufficient since in your newly added query, it's sorted by `batch_index desc`, there won't be too many iterations before finding the first finalized batch.

File: bridge-history-api/internal/logic/l1_event_parser.go [hunk 2]
```diff
@@ -270,12 +271,17 @@ func (e *L1EventParser) ParseL1BatchEventLogs(ctx context.Context, logs []types.
...
-				BatchStatus:   int(btypes.BatchStatusTypeFinalized),
-				BatchIndex:    event.BatchIndex.Uint64(),
-				BatchHash:     event.BatchHash.String(),
-				L1BlockNumber: vlog.BlockNumber,
-			})
+			var startBatchIndex uint64 = 0
+			if lastFinalizedBatchIndex, ok := f.getLastFinalizedBatchIndex(); ok {
+				startBatchIndex = lastFinalizedBatchIndex + 1
+			}
+			for index := startBatchIndex; index <= event.BatchIndex.Uint64(); index++ {
```
Comment (colinlyguo) [inline bridge-history-api/internal/logic/l1_event_parser.go:278]: What about changing the pattern of SQL update to <= batch index, and only keeping the last event (with max batch index)? This can reduce the number of SQL update ops.

File: bridge-history-api/internal/logic/l1_fetcher.go [hunk 3]
```diff
@@ -118,6 +121,12 @@ func NewL1FetcherLogic(cfg *config.FetcherConfig, db *gorm.DB, client *ethclient
 		parser:          NewL1EventParser(cfg, client),
 	}
 
+	index, err := f.batchEventOrm.GetLastFinalizedBatchIndex(context.TODO())
```
Comment (colinlyguo) [inline bridge-history-api/internal/logic/l1_fetcher.go]: passing a context from the parent call stack here when initializing the l1_fetcher instance?

File: bridge-history-api/internal/orm/batch_event.go [hunk 4]
```diff
@@ -124,3 +125,15 @@ func (c *BatchEvent) UpdateBatchEventStatus(ctx context.Context, batchIndex uint
...
+
+// GetLastFinalizedBatchIndex returns the last finalized batch index in db.
+func (c *BatchEvent) GetLastFinalizedBatchIndex(ctx context.Context) (uint64, error) {
+	var batche BatchEvent
+	db := c.db.WithContext(ctx)
+	db = db.Model(&BatchEvent{})
+	db = db.Select("batch_index")
+	db = db.Where("batch_status = ?", btypes.BatchStatusTypeFinalized)
+	db = db.Order("batch_index desc")
+	err := db.First(&batche).Error
```
Comment (colinlyguo) [inline bridge-history-api/internal/orm/batch_event.go:137]: suggesting handling `not found` explicitly, and returning 0 here when `not found` for readability.

File: bridge-history-api/internal/controller/fetcher/l1_fetcher.go [hunk 5]
```diff
@@ -153,6 +153,7 @@ func (c *L1MessageFetcher) fetchAndSaveEvents(confirmation uint64) {
 			return
 		}
 
+		c.l1FetcherLogic.ShiftToNextLastFinalizedBatchIndex()
```
Comment (colinlyguo) [inline bridge-history-api/internal/controller/fetcher/l1_fetcher.go:156]: I think updating the status here is enough. reorg won't affect this info. pls revisit this logic a bit. @amoylan2

**Comment by amoylan2** [comment]
I will open another PR with target branch "develop".
If the other is merged this can be closed anytime

File: bridge-history-api/internal/logic/l1_fetcher.go [hunk 6]
```diff
@@ -45,6 +45,9 @@ type L1FetcherLogic struct {
 	crossMessageOrm *orm.CrossMessage
 	batchEventOrm   *orm.BatchEvent
 
+	lastFinalizedBatchIndex            *uint64
+	nextPendingLastFinalizedBatchIndex uint64
```
Comment (colinlyguo) [inline bridge-history-api/internal/logic/l1_fetcher.go:49]: is this design like pre-commit, then commit? could these two variables be merged into one?

**Review by colinlyguo — APPROVED** [review]

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 6 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_fetcher.go:49]: `lastFinalizedBatchIndex` should be updated only after the events are inserted or updated to the db in which situation the context info of new value is lost.
So I add another value as a temporary store media

File: bridge-history-api/internal/controller/fetcher/l1_fetcher.go (hunk 5 above)
Comment (amoylan2) [reply bridge-history-api/internal/controller/fetcher/l1_fetcher.go:156]: The logic in line 156 is intended to make the update after the `L1InsertOrUpdate` method, where `batch_event_v2` table is updated.
I think it has nothing to do with the reorg thing

**Comment by amoylan2** [comment]
> This feature can be directly merged into the develop branch since it's backward compatible, we can deploy it first.
> 
> For discussion about: _One thing need caution: "GetLastFinalizedBatchIndex" from db can not use any index and it's not suitable to add such index since there will be almost 100% batches are all finalized as time goes. Do we need add a key-value mode cache to save the "last finalized batch index"? it can be a small kv table in postgres or using redis.._
> 
> this can be mitigated by the "composite index" as follows:
> 
> ```
> CREATE INDEX IF NOT EXISTS idx_be_batch_index_batch_status ON batch_event_v2 (batch_index, batch_status);
> ```
> 
> but I think using `batch_index` is sufficient since in your newly added query, it's sorted by `batch_index desc`, there won't be too many iterations before finding the first finalized batch.

shall we discuss in new pr ? or discuss here first

File: bridge-history-api/internal/logic/l1_event_parser.go (hunk 2 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: I'm a little confused of your comment here. 
The code snippet has nothing to do with any sql action. The `getLastFinalizedBatchIndex` method is already a local cache.
The only sql action is needed when initializing the `L1FetcherLogic` struct

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 3 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_fetcher.go]: ok, will do that

**Comment by colinlyguo** [comment]
> > This feature can be directly merged into the develop branch since it's backward compatible, we can deploy it first.
> > For discussion about: _One thing need caution: "GetLastFinalizedBatchIndex" from db can not use any index and it's not suitable to add such index since there will be almost 100% batches are all finalized as time goes. Do we need add a key-value mode cache to save the "last finalized batch index"? it can be a small kv table in postgres or using redis.._
> > this can be mitigated by the "composite index" as follows:
> > ```
> > CREATE INDEX IF NOT EXISTS idx_be_batch_index_batch_status ON batch_event_v2 (batch_index, batch_status);
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > but I think using `batch_index` is sufficient since in your newly added query, it's sorted by `batch_index desc`, there won't be too many iterations before finding the first finalized batch.
> 
> shall we discuss in new pr ? or discuss here first

both LGTM.

File: bridge-history-api/internal/orm/batch_event.go (hunk 4 above)
Comment (amoylan2) [reply bridge-history-api/internal/orm/batch_event.go:137]: I have dealt with the `not found` error at the caller of this method.
Except the `not found` error, there may other error types which will are not easy to handle in `inner` method.
In this case, the caller will panic (log.Crit) when encountering other errors since there seems no better way to handle it..
And it would be a little weird if the `inner` function only handles `not found` and return other type errors.
In conclusion I think this `inner` function should just return error and let the caller to deal with it.

File: bridge-history-api/internal/logic/l1_event_parser.go (hunk 2 above)
Comment (colinlyguo) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: I mean we can update the SQL like:
```
		case btypes.BatchStatusTypeFinalized:
			// After darwin, FinalizeBatch event signals a range of batches are finalized,
			// thus losing the batch hash info. Meanwhile, only batch_index is enough to find the target batch when finalizing batches.
			db = db.Where("batch_index <= ?", l1BatchEvent.BatchIndex)
			db = db.Where("batch_status != ?", btypes.BatchStatusTypeFinalized)
			updateFields["batch_status"] = btypes.BatchStatusTypeFinalized
			if err := db.Updates(updateFields).Error; err != nil {
				return fmt.Errorf("failed to update batch event, error: %w", err)
			}
```

File: bridge-history-api/internal/logic/l1_fetcher.go [hunk 7]
```diff
@@ -118,6 +121,16 @@ func NewL1FetcherLogic(cfg *config.FetcherConfig, db *gorm.DB, client *ethclient
 		parser:          NewL1EventParser(cfg, client),
 	}
 
+	index, err := f.batchEventOrm.GetLastFinalizedBatchIndex(context.TODO())
+	if err == gorm.ErrRecordNotFound {
+		// do nothing, leaving lastFinalizedBatchIndex to be nil
+		log.Warn("all bacthes are non-finalized, this should happen only once")
+	} else if err != nil {
+		log.Crit("failed to get last finalized bacth index, fatal db error", "err", err.Error())
```
Comment (colinlyguo) [inline bridge-history-api/internal/logic/l1_fetcher.go:129]: we can throw the other types of errors and exit this round instead of panicking here. some could succeed after retrying.

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 6 above)
Comment (colinlyguo) [reply bridge-history-api/internal/logic/l1_fetcher.go:49]: oh I see. looks like needing at least two variables to prevent early return before flushing new height to `lastFinalizedBatchIndex`.

File: bridge-history-api/internal/logic/l1_event_parser.go (hunk 2 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: got it, let me check

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 7 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_fetcher.go:129]: let me check it

File: bridge-history-api/internal/logic/l1_event_parser.go (hunk 2 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: literally one BatchEvent represents a single record in the database which relates to a single batch.
I think it's better to keep the BatchEvent array compared to change the BatchEvent meaning.

When it comes to the db operations, it can be collected the boundaries of batch indexes before doing the update. Besides, doing a range update with `<=` in a for loop is not efficient anyway.

Comment (colinlyguo) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: Doing a range update with <= in a for loop is not efficient anyway.
because after Darwin, there may be several batches per update, if we use `batch_index <= a && batch_status != finalized`, this can reduce the number of RPCs.
yeah. it's low frequency compared with cross-chain messages updates, but previously this was a very useful optimization (we merged the one-by-one read queries into one, reducing rpc latencies).

Comment (colinlyguo) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: _Doing a range update with <= in a for loop is not efficient anyway._
because after Darwin, there may be several batches per update, if we use `batch_index <= a && batch_status != finalized`, this can reduce the number of RPCs.
yeah. it's low frequency compared with cross-chain messages updates, but previously this was a very useful optimization (we merged the one-by-one read queries into one, reducing rpc latencies).

the new approach:
1. only store the FinalizeBatch event (one time), instead of splitting it into multiple events.
2. update db once.

Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_event_parser.go:278]: I have a new implementation, pls have a look

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 3 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_fetcher.go]: job is done

File: bridge-history-api/internal/logic/l1_fetcher.go [hunk 8]
```diff
@@ -45,11 +45,14 @@ type L1FetcherLogic struct {
 	crossMessageOrm *orm.CrossMessage
 	batchEventOrm   *orm.BatchEvent
 
+	lastFinalizedBatchIndex            *uint64
+	nextPendingLastFinalizedBatchIndex uint64
```
Comment (colinlyguo) [inline bridge-history-api/internal/logic/l1_fetcher.go:49]: we can remove them if using `batch_index <= x && batch_status != finalized` to update db.

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 7 above)
Comment (amoylan2) [reply bridge-history-api/internal/logic/l1_fetcher.go:129]: it's in the process initializing stage, so let's panic it

File: bridge-history-api/internal/logic/l1_fetcher.go (hunk 8 above)
Comment (colinlyguo) [reply bridge-history-api/internal/logic/l1_fetcher.go:49]: finalization can be treated as deterministic.
